Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+        byte[] ValueFromSnapshot(DocumentSnapshot snapshot)
+        {
+            if (!snapshot.Exists)
+            {
+                return null;
+            }
+            CacheDoc doc = snapshot.ConvertTo<CacheDoc>();
+            var now = DateTime.UtcNow;
+            if (doc.AbsoluteExpiration.HasValue &&
+                doc.AbsoluteExpiration.Value < now)
+            {
+                return null;
+            }
+            var slidingExpiration = doc.LastRefresh.GetValueOrDefault()
+                + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.GetValueOrDefault());
+            if (slidingExpiration < now)
+            {
+                return null;
+            }
+            return doc.Value;
+        }
+        void IDistributedCache.Refresh(string key)
+        {
+            try
+            {
+                _cacheEntries.Document(key)
+                    .UpdateAsync("LastRefresh", DateTime.UtcNow).Wait();
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        async Task IDistributedCache.RefreshAsync(string key, CancellationToken token)
+        {
+            try
+            {
+                await _cacheEntries.Document(key).UpdateAsync(
+                    "LastRefresh", DateTime.UtcNow, cancellationToken: token);
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        void IDistributedCache.Remove(string key) =>
+            _cacheEntries.Document(key).DeleteAsync().Wait();
+
+        Task IDistributedCache.RemoveAsync(string key, CancellationToken token) =>
+            _cacheEntries.Document(key).DeleteAsync(cancellationToken: token);
+
+        /// <summary>
+        /// Creates a firestore document that stores the cache value and its
+        /// expiration info, but does not store it in firestore.
+        /// </summary>
+        /// <param name="value">The value to be stored in the cache.</param>
+        /// <param name="options">Expiration info.</param>
+        /// <returns></returns>
+        CacheDoc MakeCacheDoc(byte[] value, DistributedCacheEntryOptions options)
+        {
+            CacheDoc doc = new CacheDoc()
+            {
+                LastRefresh = DateTime.UtcNow,
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds =

```

Comment:
Unwrap? (We typically aim for ~120 character line limit - 80 just ends up with far too much wrapping without any benefits.